opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateType.java

```diff
@@ -198,38 +218,73 @@ public List<DateFormatter> getTimeNamedFormatters() {
         })
         .map(DateFormatter::forPattern).collect(Collectors.toList());
   }
 
-  private ExprCoreType getExprTypeFromFormatString(String formatString) {
-    if (formatString.isEmpty()) {
-      // FOLLOW-UP: check the default formatter - and set it here instead
-      // of assuming that the default is always a timestamp
-      return TIMESTAMP;
+  private ExprCoreType getExprTypeFromCustomFormatString(String formatString) {
+    List<String> formats = Arrays.asList(formatString.split("\\|\\|"));
+    boolean isDate = false;
+    boolean isTime = false;
+
+    try {
+      // Dates will always have "year" in the formatter string
+      isDate = formats.stream()
+          .anyMatch(format -> DateTimeFormatter.ofPattern(format).toString().toLowerCase().contains("year"));
+    } catch (IllegalArgumentException ignore) {
+    }
+    try {
+      // Times will always have "hour" in the formatter string
+      isTime = formats.stream()
+          .anyMatch(format -> DateTimeFormatter.ofPattern(format).toString().toLowerCase().contains("hour"));
+    } catch (IllegalArgumentException ignore) {
     }
 
-    List<DateFormatter> namedFormatters = getAllNamedFormatters();
+    if (isDate && !isTime) {
+      return DATE;
+    }
+    if (isTime && !isDate) {
+      return TIME;
+    }
 
-    if (namedFormatters.isEmpty()) {
+    return TIMESTAMP;
+  }
+
+  private ExprCoreType getExprTypeFromFormatString(String formatString) {
+    List<DateFormatter> timeFormatters = getTimeNamedFormatters();
+    List<DateFormatter> dateFormatters = getDateNamedFormatters();
+    List<DateFormatter> customFormatters = getAllCustomFormatters();
+    List<DateFormatter> numericFormatters = getNumericNamedFormatters();
+
+    if (formatString.isEmpty()
+        || (!timeFormatters.isEmpty() && !dateFormatters.isEmpty())
+        || !numericFormatters.isEmpty()) {
+      // FOLLOW-UP: check the default formatter - and set it here instead
+      // of assuming that the default is always a timestamp
       return TIMESTAMP;
     }
 
-    if (!getAllCustomFormatters().isEmpty()) {
-      // FOLLOW-UP: support custom format in <issue#>
-      return TIMESTAMP;
+    if (!customFormatters.isEmpty()) {
+      ExprCoreType customFormatType = getExprTypeFromCustomFormatString(formatString);
+      if (!timeFormatters.isEmpty() && customFormatType == TIME) {
+        return TIME;
+      }
+      if (!dateFormatters.isEmpty() && customFormatType == DATE) {
+        return DATE;
+      }
+      return customFormatType;
     }
 
     // if there is nothing in the dateformatter that accepts a year/month/day, then
     // we can assume the type is strictly a Time object
-    if (namedFormatters.size() == getTimeNamedFormatters().size()) {
+    if (!timeFormatters.isEmpty() && dateFormatters.isEmpty()) {
       return TIME;
     }
 
     // if there is nothing in the dateformatter that accepts a hour/minute/second, then
```

Comment:
```suggestion
    // if there is nothing in the dateformatter that accepts an hour/minute/second, then
```

Tiny grammar nitpick